### ISSUE TITLE
MovingAverageEarlyStopping

### ISSUE THREAD

**Issue by nikolaosmparoutis** [keras-team/tf-keras#180]
If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**System information**.

TensorFlow version (you are using): **v2.12.0**
Are you willing to contribute it (Yes/No) : **Yes**

**Describe the feature and the current behavior/state**.
The requested feature is a custom early stopping callback for TensorFlow Keras models called `EarlyStoppingMovingAverage`. This callback extends the built-in `EarlyStopping` class and adds functionality to monitor the moving average (either simple moving average or exponential moving average) of a specified metric instead of the raw metric values.
The current behavior of the built-in `EarlyStopping` callback is to monitor a single metric (e.g., validation loss) and stop training when the metric does not improve for a specified number of consecutive epochs (patience). However, this approach can be sensitive to noisy or fluctuating metric values, leading to premature stopping.
The custom `EarlyStoppingMovingAverage` callback addresses this issue by monitoring the moving average of the metric, providing a more stable and reliable stopping criterion. This feature is needed to prevent premature stopping and improve model performance, especially in cases where the monitored metric values are noisy or fluctuate significantly during training. The use-case for this feature includes training complex models on challenging datasets where the raw metric values may not provide a clear indication of the model's performance trend.

**Will this change the current api? How?** This custom early stopping feature, `EarlyStoppingMovingAverage`, will not change the current TensorFlow Keras API directly, as it is a user-defined extension of the built-in `EarlyStopping` callback. It built to provide an additional option for users who want to implement early stopping based on the moving average of a monitored metric.
Like we did instead: 
`from tensorflow.keras.callbacks import EarlyStopping`
Now also can do:
`from tensorflow.keras.callbacks import EarlyStoppingMovingAverage`  

**Who will benefit from this feature?**
This custom early stopping feature will benefit users working with TensorFlow Keras models, particularly when the monitored metric values are noisy or fluctuate significantly during training. It can help prevent premature stopping and lead to better model performance, making it useful for researchers, data scientists, and machine learning engineers.
When datasets have many flactuations during training and validation loss does many spikes the common early stoping may stop too early the training if finds a big spike too early. Also if the training goes too far in training epochs when increase the patience could be better to stop in an earlier global or to sub-global minimum point.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): **yes**
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

It extends the built-in `EarlyStopping` class from `tensorflow.keras.callbacks` and adds functionality to monitor the moving average of a specified metric (either simple moving average or exponential moving average) instead of the raw metric values.
The `EarlyStoppingMovingAverage` class takes additional parameters like `window_size` and `moving_average_type` to control the moving average calculation. The moving average is calculated in the `_calculate_sma` and `_calculate_ema` methods, and the improvement is checked in the `_is_improvement_sma` and `_is_improvement_ema` methods.
The custom early stopping callback is designed to be more robust against noisy metric values and can help prevent premature stopping in cases where the raw metric values fluctuate significantly. By monitoring the moving average of the metric, the callback can better identify the general trend and make more informed decisions about when to stop training.
